ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java

```diff
@@ -332,71 +343,96 @@ private PreparedStatement buildSingleQueryAst(
-          ).cloneWithTimeKeyColumns());
-      if (convertStatementToQuery) {
-        return new PreparedStatement(statementString, query);
-      } else {
+    try {
+
+      if (statement instanceof Query) {
         return new PreparedStatement(statementString, statement);
       }
 
-    } else if (statement instanceof InsertInto) {
-      final InsertInto insertInto = (InsertInto) statement;
-      if (tempMetaStoreForParser.getSource(insertInto.getTarget().getSuffix()) == null) {
-        throw new KsqlException(String.format("%s. Error: Sink, "
-                + "%s, does not exist for the INSERT INTO statement.",
-            statementString, insertInto.getTarget().getSuffix()));
+      if (statement instanceof CreateAsSelect) {
+        return buildCreateAsSelectStatement(
+            statement, statementString, tempMetaStoreForParser, convertStatementToQuery);
       }
 
-      if (tempMetaStoreForParser.getSource(insertInto.getTarget().getSuffix()).getDataSourceType()
-          != DataSource.DataSourceType.KSTREAM) {
-        throw new KsqlException(String.format("INSERT INTO can only be used to insert into a "
-                                              + "stream. %s is a table.",
-                                              insertInto.getTarget().getSuffix()));
+      if (statement instanceof InsertInto) {
+        return buildInsertIntoStatement(
+            statement, statementString, tempMetaStoreForParser, convertStatementToQuery);
       }
 
-      final QuerySpecification querySpecification =
-          (QuerySpecification) insertInto.getQuery().getQueryBody();
-
-      final Query query = addInto(
-          insertInto.getQuery(),
-          querySpecification,
-          insertInto.getTarget().getSuffix(),
-          new HashMap<>(),
-          insertInto.getPartitionByColumn(),
-          false
-      );
-
-      if (convertStatementToQuery) {
-        return new PreparedStatement(statementString, query);
-      } else {
-        return new PreparedStatement(statementString, statement);
+      if (statement instanceof DdlStatement) {
+        return buildSingleDdlStatement(
+            statement, statementString, tempMetaStore, tempMetaStoreForParser);
       }
-    } else  if (statement instanceof DdlStatement) {
-      return buildSingleDdlStatement(statement,
-                                     statementString,
-                                     tempMetaStore,
-                                     tempMetaStoreForParser);
+
+      return new PreparedStatement(statementString, statement);
+      //    } catch (final KsqlException e) {
+      //   throw e;  -> KsqlEngineTest::shouldFailIfReferenialIntegityVolated
+    } catch (final Exception e) {
+      throw new ParseFailedException(
+          "Exception while processing statement: " + e.getMessage(), statementString, e);
     }
+  }
 
-    return null;
+  private PreparedStatement buildCreateAsSelectStatement(final Statement statement,
```

Comment:
We should use a different verb for these (maybe postprocess ?). in buildMultipleQueries "build" to means to take statement text and turn it into a startable object that runs the query, and to update the metastore. Here its to do some post-processing on a parsed statement to produce the final parsed statement.

Reply:
done